neo/Network/P2P/Payloads/Transaction.cs

```diff
@@ -56,72 +59,30 @@ public UInt256 Hash
-            Script.GetVarSize() +       //Script
-            Sender.Size +               //Sender
-            sizeof(long) +              //Gas
-            sizeof(long) +              //NetworkFee
-            sizeof(uint) +              //ValidUntilBlock
-            Attributes.GetVarSize() +   //Attributes
-            Witness.Size;               //Witnesses
+        public const int HeaderSize =
+            sizeof(byte) +  //Version
+            sizeof(uint) +  //Nonce
+            20 +            //Sender
+            sizeof(long) +  //Gas
+            sizeof(long) +  //NetworkFee
+            sizeof(uint);   //ValidUntilBlock
 
-        public void CalculateFees()
-        {
-            if (Sender is null) Sender = UInt160.Zero;
-            if (Attributes is null) Attributes = new TransactionAttribute[0];
-            if (Witness is null) Witness = new Witness
-            {
-                InvocationScript = new byte[65],
-                VerificationScript = new byte[39]
-            };
-            _hash = null;
-            long consumed;
-            using (ApplicationEngine engine = ApplicationEngine.Run(Script, this))
-            {
-                if (engine.State.HasFlag(VMState.FAULT))
-                    throw new InvalidOperationException();
-                consumed = engine.GasConsumed;
-            }
-            _hash = null;
-            long d = (long)NativeContract.GAS.Factor;
-            Gas = consumed - ApplicationEngine.GasFree;
-            if (Gas <= 0)
-            {
-                Gas = 0;
-            }
-            else
-            {
-                long remainder = Gas % d;
-                if (remainder == 0) return;
-                if (remainder > 0)
-                    Gas += d - remainder;
-                else
-                    Gas -= remainder;
-            }
-            using (Snapshot snapshot = Blockchain.Singleton.GetSnapshot())
-            {
-                long feeperbyte = NativeContract.Policy.GetFeePerByte(snapshot);
-                long fee = feeperbyte * Size;
-                if (fee > NetworkFee)
-                    NetworkFee = fee;
-            }
-        }
+        public int Size => HeaderSize +
+            Attributes.GetVarSize() +   //Attributes
+            Script.GetVarSize() +       //Script
+            Witnesses.GetVarSize();     //Witnesses
 
         void ISerializable.Deserialize(BinaryReader reader)
         {
             DeserializeUnsigned(reader);
-            Witness = reader.ReadSerializable<Witness>();
+            Witnesses = reader.ReadSerializableArray<Witness>();
         }
 
         public void DeserializeUnsigned(BinaryReader reader)
         {
             Version = reader.ReadByte();
             if (Version > 0) throw new FormatException();
             Nonce = reader.ReadUInt32();
-            Script = reader.ReadVarBytes(ushort.MaxValue);
-            if (Script.Length == 0) throw new FormatException();
```

Comment:
Erik, I think that Script must be part of signed part. Othewise, how can we be sure that the invoked NEP-5 transfer is actually transfering the desired amount?

Reply:
Ok, I saw it just changed order. This is fine.